security/advancedtls/advancedtls_integration_test.go

```diff
@@ -408,3 +411,272 @@ func (s) TestEnd2End(t *testing.T) {
+			// files.
+			clientIdentityProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				CertFile:         clientCertTmp.Name(),
+				KeyFile:          clientKeyTmp.Name(),
+				IdentityInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientIdentityProvider.Close()
+			clientRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    clientTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientRootProvider.Close()
+			serverIdentityProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				CertFile:         serverCertTmp.Name(),
+				KeyFile:          serverKeyTmp.Name(),
+				IdentityInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverIdentityProvider.Close()
+			serverRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    serverTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverRootProvider.Close()
+			// Start a server and create a client using advancedtls API with Provider.
+			serverOptions := &ServerOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: serverIdentityProvider,
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: serverRootProvider,
+				},
+				RequireClientCert: true,
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				VType: CertVerification,
+			}
+			serverTLSCreds, err := NewServerCreds(serverOptions)
+			if err != nil {
+				t.Fatalf("failed to create server creds: %v", err)
+			}
+			s := grpc.NewServer(grpc.Creds(serverTLSCreds))
+			defer s.Stop()
+			lis, err := net.Listen("tcp", port)
+			if err != nil {
+				t.Fatalf("failed to listen: %v", err)
+			}
+			defer lis.Close()
+			pb.RegisterGreeterServer(s, greeterServer{})
+			go s.Serve(lis)
+			clientOptions := &ClientOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: clientIdentityProvider,
+				},
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: clientRootProvider,
+				},
+				VType: CertVerification,
+			}
+			clientTLSCreds, err := NewClientCreds(clientOptions)
+			if err != nil {
+				t.Fatalf("clientTLSCreds failed to create, error: %v", err)
+			}
+			// At initialization, the connection should be good.
+			ctx1, cancel1 := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel1()
+			conn, greetClient, err := callAndVerifyWithClientConn(ctx1, "rpc call 1", clientTLSCreds, false)
```

Comment:
Not required to be done as part of this PR. But `callAndVerifyWithClientConn` and `callAndVerify` should be changed to not accept `shouldFail`. That logic is better to be done at the test.

And in `callAndVerifyWithClientConn`, there is the following comment:
```
	// If we want the test to fail, we establish a non-blocking connection to
	// avoid it hangs and killed by the context.
```
But it looks like it is doing the opposite. And it should OK to do a blocking dial under all circumstances from the test. So, if `callAndVerifyWithClientConn` is changed to accept `...grpc.DialOption` as its last argument, callers can control what dial options are passed.

Reply:
Sure, I added some TODO(s) and will clean them up later.